File: cybership_observer/src/ned_to_enu_transformer.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Matrix3x3.h"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"

class NEDToENUPublisher : public rclcpp::Node
{


private:
    rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr ned_sub_;

    rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr enu_pub_;

    std::string frame_id_;

    void ned_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
    {
        // Transform the pose from NED to ENU frame
        geometry_msgs::msg::PoseWithCovarianceStamped enu_msg;
        enu_msg.header = msg->header;

        enu_msg.header.frame_id = frame_id_;

        // Position transformation
        enu_msg.pose.pose.position.x = msg->pose.pose.position.y;
        enu_msg.pose.pose.position.y = msg->pose.pose.position.x;
        enu_msg.pose.pose.position.z = -msg->pose.pose.position.z;

        // Orientation transformation (quaternion)
        tf2::Quaternion q_ned(
            msg->pose.pose.orientation.x,
            msg->pose.pose.orientation.y,
            msg->pose.pose.orientation.z,
            msg->pose.pose.orientation.w);


        tf2::Matrix3x3 m(q_ned);
        double yaw, pitch, roll;
        m.getRPY(yaw, pitch, roll);

        tf2::Quaternion q_enu;

        q_enu.setRPY(roll, pitch, -yaw + M_PI_2);

        enu_msg.pose.pose.orientation.x = q_enu.x();
        enu_msg.pose.pose.orientation.y = q_enu.y();
        enu_msg.pose.pose.orientation.z = q_enu.z();
        enu_msg.pose.pose.orientation.w = q_enu.w();

        // Copy the covariance matrix as it is (assuming it remains the same)
        enu_msg.pose.covariance = msg->pose.covariance;

        // Publish the transformed message
        enu_pub_->publish(enu_msg);
    }

public:
    NEDToENUPublisher()
        : Node("ned_to_enu_publisher")
    {
        // // Subscriber to the NED frame PoseWithCovarianceStamped messages
        ned_sub_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
            "in_pose", 10, std::bind(&NEDToENUPublisher::ned_callback, this, std::placeholders::_1));

        // Publisher for the transformed ENU frame PoseWithCovarianceStamped messages
        enu_pub_ = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("out_pose", 10);

        // Read `frame_id` parameter
        this->declare_parameter<std::string>("frame_id", "world");
        this->get_parameter("frame_id", frame_id_);
    }

};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<NEDToENUPublisher>());
    rclcpp::shutdown();
    return 0;
}

File: cybership_observer/nodes/ned_to_enu_transformer.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseWithCovarianceStamped
import tf_transformations
import numpy as np

class NEDToENUPublisher(Node):

    def __init__(self):
        super().__init__('ned_to_enu_publisher')

        # Declare and get the frame_id parameter
        self.declare_parameter('frame_id', 'world')
        self.frame_id = self.get_parameter(
            'frame_id').get_parameter_value().string_value

        # Subscriber to the NED frame PoseWithCovarianceStamped messages
        self.ned_sub = self.create_subscription(
            PoseWithCovarianceStamped,
            'in_pose',
            self.ned_callback,
            10
        )

        # Publisher for the transformed ENU frame PoseWithCovarianceStamped messages
        self.enu_pub = self.create_publisher(
            PoseWithCovarianceStamped,
            'out_pose',
            10
        )

    def ned_callback(self, msg):
        # Transform the pose from NED to ENU frame
        enu_msg = PoseWithCovarianceStamped()
        enu_msg.header = msg.header
        enu_msg.header.frame_id = self.frame_id  # Set the frame_id parameter

        # Position transformation
        enu_msg.pose.pose.position.x = msg.pose.pose.position.y
        enu_msg.pose.pose.position.y = msg.pose.pose.position.x
        enu_msg.pose.pose.position.z = -msg.pose.pose.position.z

        # Orientation transformation (quaternion)
        q_ned = [
            msg.pose.pose.orientation.x,
            msg.pose.pose.orientation.y,
            msg.pose.pose.orientation.z,
            msg.pose.pose.orientation.w
        ]

        roll, pitch, yaw = tf_transformations.euler_from_quaternion(
            q_ned, axes='sxyz')  # Check the orientation in Euler angles

        q_enu = tf_transformations.quaternion_from_euler(
            roll, pitch, -yaw + np.pi / 2.0, axes='sxyz')

        enu_msg.pose.pose.orientation.x = q_enu[0]
        enu_msg.pose.pose.orientation.y = q_enu[1]
        enu_msg.pose.pose.orientation.z = q_enu[2]
        enu_msg.pose.pose.orientation.w = q_enu[3]

        # Copy the covariance matrix as it is (assuming it remains the same)
        enu_msg.pose.covariance = msg.pose.covariance

        # Publish the transformed message
        self.enu_pub.publish(enu_msg)


def main(args=None):
    rclpy.init(args=args)
    node = NEDToENUPublisher()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
